composer.json

```diff
@@ -3,9 +3,10 @@
     "description": "Eloquent syntax for DynamoDB",
     "keywords": ["laravel", "dynamodb", "aws"],
     "require": {
+        "php": "^7.1.3",
```

Comment:
What would happen if people are using Laravel 5.2 with older PHP version and they want to upgrade this library for bug fixes?

Reply:
The semantic versioning in the composer.json *should* stop them from using the newer version. Hence the suggestion of a new major version to stop any accidental breaking changes.

Reply:
I'd like them to be able to use the newer version of this library though because there are some bug fixes. We can improve on versioning later. Why do you think people using older PHP version should not upgrade this library?

Reply:
Anyone that is using a _currently supported_ version of PHP will be able to use this version. Laravel 5.8 requires PHP 7.1 and above - you can’t support both Laravel 5.8 and PHP versions less than 7.1.

Reply:
I'm talking about people who are using Laravel 5.2 for example

Reply:
Laravel 5.2 is no longer supported by Laravel, and hasn’t been since 2016. In my opinion, supporting a newer version of the framework should supersede one that’s been unsupported for 3 years.

Reply:
If there are bugfixes that should be made available to people using an older version of Laravel, they could be included in both the next patch release and the next minor release. You could continue to support Laravel 5.2 in release v4.11.3 (and include the bugfixes), while dropping support for older versions of Laravel and maintaining compatibility with the latest Laravel releases in v.4.12.0 or v5.0.0 of this library. But I agree with @footballencarta that it makes no sense to hold out on supporting Laravel 5.8 because it might exclude people using a framework version that hasn't been supported for 3 years now.

Reply:
I don't think everyone would have the option to use newer version of Laravel. Tests will faill of course, but so far it should still work for those people. I don't want to support newer version of Laravel and abandon immediately all the older versions. That would need careful planning and I don't feel it should be done in this PR.

Reply:
Sounds like you need to start a new v5 release with this PR, while maintaining v4 with bugfixes only for older versions of Laravel. Old Laravel users can easily restrict this package to `^4.0` in their composer.json file. Right?

I understand the concerns about those older Laravel users. That said, there are a lot of us that _really_ need to upgrade to Laravel 5.8 and this is a blocker.

Reply:
@footballencarta Since Laravel enforces a minimum PHP version, maybe it's not necessary to add it in this package as well? Anyone who is trying to install this package in Laravel 5.8 will already have dealt with Laravel's PHP 7.1 requirement.

Reply:
@footballencarta You could remove the PHP version requirement here since Laravel will have already enforced that.

However, I see you've also added a bunch of PHP 7.1 specific code, like function return types. IMHO these are unnecessary and preventing this from getting merged quickly.

I would suggest backing off the PHP requirement (anyone installing this in Laravel 5.8 will have already dealt with Laravel's PHP 7.1 requirement) and removing your function return types.

A simple composer.json change to allow illuminate/support and illuminate/database at 5.8 will enable this package to be installed in Laravel 5.8 without breaking older versions as well. Seems like a simple and quick fix. Later if @baopham decides to have a separate major release for new versions of PHP & Laravel, PHP 7.1 specific function return types could be added.

...unless I'm missing something?

Reply:
The return types are enforced by the version of PHPUnit that ships with Laravel 5.8 - in that it expects a return type to be added, otherwise it fails the entire test set for invalid code (PHPUnit 8+).

I understand the issue with supporting older versions - which is why all along I’ve suggest merging this into a new major version, as it contains breaking changes. This will allow the package to be developed to keep pace with Laravel’s standards, while also backporting big fixes to “legacy” versions of Laravel / PHP if needed.

Honestly, we’re in 2019 now, and PHP 7.1 has been around since the end of 2016. I get some people can’t upgrade for whatever reasons (hell we have some guys where I work that refuse to switch from 5.6) - but making this as a new major version supporting the next generation of Laravel applications is better than trying to maintain backwards compatibility with _everything_ and eventually being left vastly far behind.

Reply:
https://github.com/laravel/laravel/blob/master/composer.json#L22

https://github.com/laravel/framework/blob/master/composer.json#L88

I'm seeing PHPUnit 7.5 in laravel/laravel master, and either 7.5 or 8 in laravel/framework. Are you sure PHPUnit 8 is a hard requirement? Not seeing it.

Reply:
If you follow @jszobody's suggestion, then we can go ahead with this PR. I don't think forcing the PHP version should be in this PR. Why do you need this line to unblock you?